hail/src/main/scala/is/hail/annotations/aggregators/RegionValueApproxCDFAggregator.scala

```diff
@@ -0,0 +1,706 @@
+package is.hail.annotations.aggregators
+
+import is.hail.annotations._
+import is.hail.expr.types.virtual._
+import is.hail.utils.ArrayBuilder
+import org.apache.spark.sql.Row
+import net.sourceforge.jdistlib.rng.MersenneTwister
+
+import scala.reflect.ClassTag
+import scala.math.Ordering
+
+object ApproxCDFHelper {
+  implicit object IntHelper extends ApproxCDFHelper[Int] {
+    def compare(x: Int, y: Int): Int = java.lang.Integer.compare(x, y)
+    def lt(x: Int, y: Int): Boolean = x < y
+    def sort(a: Array[Int], begin: Int, end: Int) {
+      java.util.Arrays.sort(a, begin, end)
+    }
+    def dummyValue: Int = 0
+    val hailType = TInt32()
+  }
+  implicit object LongHelper extends ApproxCDFHelper[Long] {
+    def compare(x: Long, y: Long): Int = java.lang.Long.compare(x, y)
+    def lt(x: Long, y: Long): Boolean = x < y
+    def sort(a: Array[Long], begin: Int, end: Int) {
+      java.util.Arrays.sort(a, begin, end)
+    }
+    def dummyValue: Long = 0
+    val hailType = TInt64()
+  }
+  implicit object FloatHelper extends ApproxCDFHelper[Float] {
+    def compare(x: Float, y: Float): Int = java.lang.Float.compare(x, y)
+    def lt(x: Float, y: Float): Boolean = x < y
+    def sort(a: Array[Float], begin: Int, end: Int) {
+      java.util.Arrays.sort(a, begin, end)
+    }
+    def dummyValue: Float = 0
+    val hailType = TFloat32()
+  }
+  implicit object DoubleHelper extends ApproxCDFHelper[Double] {
+    def compare(x: Double, y: Double): Int = java.lang.Double.compare(x, y)
+    def lt(x: Double, y: Double): Boolean = x < y
+    def sort(a: Array[Double], begin: Int, end: Int) {
+      java.util.Arrays.sort(a, begin, end)
+    }
+    def dummyValue: Double = 0
+    val hailType = TFloat64()
+  }
+}
+
+abstract class ApproxCDFHelper[@specialized(Int, Long, Float, Double) T] extends Serializable {
+  private val rand = new java.util.Random()
+
+  val hailType: Type
+
+  def dummyValue: T
+
+  def compare(x: T, y: T): Int
+  def lt(x: T, y: T): Boolean
+
+  def min(x: T, y: T): T = if (lt(x, y)) x else y
+  def max(x: T, y: T): T = if (lt(x, y)) y else x
+
+  def sort(a: Array[T], begin: Int, end: Int): Unit
+
+  def merge(
+    left: Array[T], lStart: Int, lEnd: Int,
+    right: Array[T], rStart: Int, rEnd: Int,
+    out: Array[T], outStart: Int
```

Comment:
Hmm, I think this has this prerequisite, right?
```scala
assert(left != out || outStart >= lEnd)
assert(right != out || outStart >= rEnd)
```

I think there's actually a slightly looser prerequisite involving the max of their sizes, but this should be sufficient for our purposes.

Reply:
This also is most commonly called with `left == right == out`, and with the output range overlapping one of the input ranges. I can work out the exact preconditions.